In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [ ]:
anime=pd.read_csv("/content/drive/MyDrive/Assignments/Recommendation system/anime.csv")
anime.head()

anime_id                              name  \
0     32281                    Kimi no Na wa.   
1      5114  Fullmetal Alchemist: Brotherhood   
2     28977                          Gintama°   
3      9253                       Steins;Gate   
4      9969                     Gintama&#039;   

                                               genre   type episodes  rating  \
0               Drama, Romance, School, Supernatural  Movie        1    9.37   
1  Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64    9.26   
2  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.25   
3                                   Sci-Fi, Thriller     TV       24    9.17   
4  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.16   

   members  
0   200630  
1   793665  
2   114262  
3   673572  
4   151266


**Data Description:**

- Unique ID of each anime.
- Anime title.
- Anime broadcast type, such as TV, OVA, etc.
- anime genre.
- The number of episodes of each anime.
- The average rating for each anime compared to the number of users who gave ratings.
- Number of community members for each anime.


In [ ]:
anime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


episodes are not continous data, but they are numerical data

In [ ]:
anime.isnull().sum()

anime_id      0
name          0
genre        62
type         25
episodes      0
rating      230
members       0
dtype: int64

- there are null values in attributes like genre, type, and rating

In [ ]:
anime.describe()

anime_id        rating       members
count  12294.000000  12064.000000  1.229400e+04
mean   14058.221653      6.473902  1.807134e+04
std    11455.294701      1.026746  5.482068e+04
min        1.000000      1.670000  5.000000e+00
25%     3484.250000      5.880000  2.250000e+02
50%    10260.500000      6.570000  1.550000e+03
75%    24794.500000      7.180000  9.437000e+03
max    34527.000000     10.000000  1.013917e+06

In [ ]:
anime.shape

(12294, 7)

In [ ]:
print(len(anime["name"].unique()))

12292


- there are 12292 unique values in the dataset, that is 12992 unique anime names in the dataset


In [ ]:
print(len(anime["genre"].unique()))

3265


In [ ]:
anime.sort_values("anime_id")

anime_id                                               name  \
22            1                                       Cowboy Bebop   
152           5                    Cowboy Bebop: Tengoku no Tobira   
214           6                                             Trigun   
2095          7                                 Witch Hunter Robin   
3159          8                             Beet the Vandel Buster   
...         ...                                                ...   
2726      34514                                Pokemon Generations   
9586      34519                Mobile Suit Gakuen: G-Reco Koushien   
11104     34522                          Wake Up, Girls! Shin Shou   
10952     34525                                  Centaur no Nayami   
10997     34527  Gou-chan. Moko to Chinjuu no Mori no Nakama-tachi   

                                                   genre     type episodes  \
22       Action, Adventure, Comedy, Drama, Sci-Fi, Space       TV       26   
152                Action, Drama, Mystery, Sci-Fi, Space    Movie        1   
214                               Action, Comedy, Sci-Fi       TV       26   
2095   Action, Drama, Magic, Mystery, Police, Superna...       TV       26   
3159           Adventure, Fantasy, Shounen, Supernatural       TV       52   
...                                                  ...      ...      ...   
2726              Action, Adventure, Fantasy, Game, Kids      ONA       18   
9586                                              Comedy  Special        9   
11104                                       Drama, Music       TV  Unknown   
10952       Comedy, Fantasy, Slice of Life, Supernatural       TV  Unknown   
10997                                    Adventure, Kids    Movie        1   

       rating  members  
22       8.82   486824  
152      8.40   137636  
214      8.32   283069  
2095     7.36    64905  
3159     7.06     9848  
...       ...      ...  
2726     7.21      295  
9586     5.67       94  
11104     NaN      381  
10952     NaN      108  
10997     NaN        5  

[12294 rows x 7 columns]

In [ ]:
len(anime["anime_id"].unique())

12294

In [ ]:
from sklearn.preprocessing import OneHotEncoder, MinMaxScaler
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.model_selection import train_test_split
from sklearn.metrics import precision_score, recall_score, f1_score

In [ ]:
anime["rating"] = anime["rating"].astype(float)

In [ ]:
anime["rating"].fillna(anime["rating"].median(),inplace = True)

In [ ]:
anime.isnull().sum()

anime_id     0
name         0
genre       62
type        25
episodes     0
rating       0
members      0
dtype: int64

In [ ]:
anime.dropna(inplace=True)

In [ ]:
anime.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12210 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12210 non-null  int64  
 1   name      12210 non-null  object 
 2   genre     12210 non-null  object 
 3   type      12210 non-null  object 
 4   episodes  12210 non-null  object 
 5   rating    12210 non-null  float64
 6   members   12210 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 763.1+ KB


- so all null values are gone and know we can work on the recommendation system

In [ ]:
anime1 = anime.pivot(index='anime_id',columns='name',values='rating')
anime1

name      &quot;0&quot;  \
anime_id                  
1                   NaN   
5                   NaN   
6                   NaN   
7                   NaN   
8                   NaN   
...                 ...   
34514               NaN   
34519               NaN   
34522               NaN   
34525               NaN   
34527               NaN   

name      &quot;Aesop&quot; no Ohanashi yori: Ushi to Kaeru, Yokubatta Inu  \
anime_id                                                                     
1                                                       NaN                  
5                                                       NaN                  
6                                                       NaN                  
7                                                       NaN                  
8                                                       NaN                  
...                                                     ...                  
34514                                                   NaN                  
34519                                                   NaN                  
34522                                                   NaN                  
34525                                                   NaN                  
34527                                                   NaN                  

name      &quot;Bungaku Shoujo&quot; Kyou no Oyatsu: Hatsukoi  \
anime_id                                                        
1                                                       NaN     
5                                                       NaN     
6                                                       NaN     
7                                                       NaN     
8                                                       NaN     
...                                                     ...     
34514                                                   NaN     
34519                                                   NaN     
34522                                                   NaN     
34525                                                   NaN     
34527                                                   NaN     

name      &quot;Bungaku Shoujo&quot; Memoire  \
anime_id                                       
1                                        NaN   
5                                        NaN   
6                                        NaN   
7                                        NaN   
8                                        NaN   
...                                      ...   
34514                                    NaN   
34519                                    NaN   
34522                                    NaN   
34525                                    NaN   
34527                                    NaN   

name      &quot;Bungaku Shoujo&quot; Movie  &quot;Eiji&quot;  \
anime_id                                                       
1                                      NaN               NaN   
5                                      NaN               NaN   
6                                      NaN               NaN   
7                                      NaN               NaN   
8                                      NaN               NaN   
...                                    ...               ...   
34514                                  NaN               NaN   
34519                                  NaN               NaN   
34522                                  NaN               NaN   
34525                                  NaN               NaN   
34527                                  NaN               NaN   

name      &quot;Eiyuu&quot; Kaitai  .hack//G.U. Returner  .hack//G.U. Trilogy  \
anime_id                                                                        
1                              NaN                   NaN                  NaN   
5                              NaN                   NaN                  NaN   
6    

In [ ]:
anime1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12210 entries, 1 to 34527
Columns: 12208 entries, &quot;0&quot; to ◯
dtypes: float64(12208)
memory usage: 1.1 GB


In [ ]:
anime1.fillna(0,inplace=True)

In [ ]:
anime1

name      &quot;0&quot;  \
anime_id                  
1                   0.0   
5                   0.0   
6                   0.0   
7                   0.0   
8                   0.0   
...                 ...   
34514               0.0   
34519               0.0   
34522               0.0   
34525               0.0   
34527               0.0   

name      &quot;Aesop&quot; no Ohanashi yori: Ushi to Kaeru, Yokubatta Inu  \
anime_id                                                                     
1                                                       0.0                  
5                                                       0.0                  
6                                                       0.0                  
7                                                       0.0                  
8                                                       0.0                  
...                                                     ...                  
34514                                                   0.0                  
34519                                                   0.0                  
34522                                                   0.0                  
34525                                                   0.0                  
34527                                                   0.0                  

name      &quot;Bungaku Shoujo&quot; Kyou no Oyatsu: Hatsukoi  \
anime_id                                                        
1                                                       0.0     
5                                                       0.0     
6                                                       0.0     
7                                                       0.0     
8                                                       0.0     
...                                                     ...     
34514                                                   0.0     
34519                                                   0.0     
34522                                                   0.0     
34525                                                   0.0     
34527                                                   0.0     

name      &quot;Bungaku Shoujo&quot; Memoire  \
anime_id                                       
1                                        0.0   
5                                        0.0   
6                                        0.0   
7                                        0.0   
8                                        0.0   
...                                      ...   
34514                                    0.0   
34519                                    0.0   
34522                                    0.0   
34525                                    0.0   
34527                                    0.0   

name      &quot;Bungaku Shoujo&quot; Movie  &quot;Eiji&quot;  \
anime_id                                                       
1                                      0.0               0.0   
5                                      0.0               0.0   
6                                      0.0               0.0   
7                                      0.0               0.0   
8                                      0.0               0.0   
...                                    ...               ...   
34514                                  0.0               0.0   
34519                                  0.0               0.0   
34522                                  0.0               0.0   
34525                                  0.0               0.0   
34527                                  0.0               0.0   

name      &quot;Eiyuu&quot; Kaitai  .hack//G.U. Returner  .hack//G.U. Trilogy  \
anime_id                                                                        
1                              0.0                   0.0                  0.0   
5                              0.0                   0.0                  0.0   
6    

In [ ]:
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine,correlation

##Cosine similarity using name

In [ ]:
anime_sim = 1 - pairwise_distances( anime1.values,metric='cosine')

In [ ]:
anime_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [ ]:
anime_sim_df =pd.DataFrame(anime_sim)
anime_sim_df

0      1      2      3      4      5      6      7      8      9      \
0        1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1        0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2        0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3        0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0   
4        0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
12205    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
12206    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
12207    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
12208    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
12209    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

       ...  12200  12201  12202  12203  12204  12205  12206  12207  12208  \
0      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
12205  ...    0.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0   
12206  ...    0.0    0.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0   
12207  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1.0    0.0   
12208  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1.0   
12209  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

       12209  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
...      ...  
12205    0.0  
12206    0.0  
12207    0.0  
12208    0.0  
12209    1.0  

[12210 rows x 12210 columns]

In [ ]:
anime_sim_df.index= anime1.index.unique()
anime_sim_df.columns= anime1.index.unique()

In [ ]:
anime_sim_df

anime_id  1      5      6      7      8      15     16     17     18     \
anime_id                                                                  
1           1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5           0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
6           0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0   
7           0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   
8           0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0   
...         ...    ...    ...    ...    ...    ...    ...    ...    ...   
34514       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34519       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34522       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34525       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34527       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  19     ...  34497  34498  34502  34503  34506  34514  34519  34522  \
anime_id         ...                                                           
1           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
6           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
7           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
8           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...         ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
34514       0.0  ...    0.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0   
34519       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    1.0    0.0   
34522       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1.0   
34525       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34527       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  34525  34527  
anime_id                
1           0.0    0.0  
5           0.0    0.0  
6           0.0    0.0  
7           0.0    0.0  
8           0.0    0.0  
...         ...    ...  
34514       0.0    0.0  
34519       0.0    0.0  
34522       0.0    0.0  
34525       1.0    0.0  
34527       0.0    1.0  

[12210 rows x 12210 columns]

In [ ]:
np.fill_diagonal(anime_sim,val=0)
anime_sim_df


anime_id  1      5      6      7      8      15     16     17     18     \
anime_id                                                                  
1           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
6           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
7           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
8           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...         ...    ...    ...    ...    ...    ...    ...    ...    ...   
34514       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34519       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34522       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34525       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34527       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  19     ...  34497  34498  34502  34503  34506  34514  34519  34522  \
anime_id         ...                                                           
1           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
6           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
7           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
8           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...         ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
34514       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34519       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34522       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34525       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34527       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  34525  34527  
anime_id                
1           0.0    0.0  
5           0.0    0.0  
6           0.0    0.0  
7           0.0    0.0  
8           0.0    0.0  
...         ...    ...  
34514       0.0    0.0  
34519       0.0    0.0  
34522       0.0    0.0  
34525       0.0    0.0  
34527       0.0    0.0  

[12210 rows x 12210 columns]

In [ ]:
anime_sim_df.idxmax(axis=1)[:10]

anime_id
1     1
5     1
6     1
7     1
8     1
15    1
16    1
17    1
18    1
19    1
dtype: int64

In [ ]:
anime_sim_df[anime_sim_df.columns[0]].max()

0.0

##Cosine similarity using Genre

In [ ]:
anime_genre= anime.pivot(index='anime_id',columns='genre',values='rating')
anime_genre

genre     Action  Action, Adventure  \
anime_id                              
1            NaN                NaN   
5            NaN                NaN   
6            NaN                NaN   
7            NaN                NaN   
8            NaN                NaN   
...          ...                ...   
34514        NaN                NaN   
34519        NaN                NaN   
34522        NaN                NaN   
34525        NaN                NaN   
34527        NaN                NaN   

genre     Action, Adventure, Cars, Comedy, Sci-Fi, Shounen  \
anime_id                                                     
1                                                      NaN   
5                                                      NaN   
6                                                      NaN   
7                                                      NaN   
8                                                      NaN   
...                                                    ...   
34514                                                  NaN   
34519                                                  NaN   
34522                                                  NaN   
34525                                                  NaN   
34527                                                  NaN   

genre     Action, Adventure, Cars, Mecha, Sci-Fi, Shounen, Sports  \
anime_id                                                            
1                                                       NaN         
5                                                       NaN         
6                                                       NaN         
7                                                       NaN         
8                                                       NaN         
...                                                     ...         
34514                                                   NaN         
34519                                                   NaN         
34522                                                   NaN         
34525                                                   NaN         
34527                                                   NaN         

genre     Action, Adventure, Cars, Sci-Fi  Action, Adventure, Comedy  \
anime_id                                                               
1                                     NaN                        NaN   
5                                     NaN                        NaN   
6                                     NaN                        NaN   
7                                     NaN                        NaN   
8                                     NaN                        NaN   
...                                   ...                        ...   
34514                                 NaN                        NaN   
34519                                 NaN                        NaN   
34522                                 NaN                        NaN   
34525                                 NaN                        NaN   
34527                                 NaN                        NaN   

genre     Action, Adventure, Comedy, Demons, Drama, Ecchi, Horror, Mystery, Romance, Sci-Fi  \
anime_id                                                                                      
1                                                       NaN                                   
5                                                       NaN                                   
6                                                       NaN                                   
7                                                       NaN                                   
8                                                       NaN                                   
...                                                     ...                                   
34514                                                   NaN                               

In [ ]:
anime_genre.fillna(0,inplace=True)
anime_genre

genre     Action  Action, Adventure  \
anime_id                              
1            0.0                0.0   
5            0.0                0.0   
6            0.0                0.0   
7            0.0                0.0   
8            0.0                0.0   
...          ...                ...   
34514        0.0                0.0   
34519        0.0                0.0   
34522        0.0                0.0   
34525        0.0                0.0   
34527        0.0                0.0   

genre     Action, Adventure, Cars, Comedy, Sci-Fi, Shounen  \
anime_id                                                     
1                                                      0.0   
5                                                      0.0   
6                                                      0.0   
7                                                      0.0   
8                                                      0.0   
...                                                    ...   
34514                                                  0.0   
34519                                                  0.0   
34522                                                  0.0   
34525                                                  0.0   
34527                                                  0.0   

genre     Action, Adventure, Cars, Mecha, Sci-Fi, Shounen, Sports  \
anime_id                                                            
1                                                       0.0         
5                                                       0.0         
6                                                       0.0         
7                                                       0.0         
8                                                       0.0         
...                                                     ...         
34514                                                   0.0         
34519                                                   0.0         
34522                                                   0.0         
34525                                                   0.0         
34527                                                   0.0         

genre     Action, Adventure, Cars, Sci-Fi  Action, Adventure, Comedy  \
anime_id                                                               
1                                     0.0                        0.0   
5                                     0.0                        0.0   
6                                     0.0                        0.0   
7                                     0.0                        0.0   
8                                     0.0                        0.0   
...                                   ...                        ...   
34514                                 0.0                        0.0   
34519                                 0.0                        0.0   
34522                                 0.0                        0.0   
34525                                 0.0                        0.0   
34527                                 0.0                        0.0   

genre     Action, Adventure, Comedy, Demons, Drama, Ecchi, Horror, Mystery, Romance, Sci-Fi  \
anime_id                                                                                      
1                                                       0.0                                   
5                                                       0.0                                   
6                                                       0.0                                   
7                                                       0.0                                   
8                                                       0.0                                   
...                                                     ...                                   
34514                                                   0.0                               

In [ ]:
anime_genre.max()

genre
Action                                                     7.83
Action, Adventure                                          8.00
Action, Adventure, Cars, Comedy, Sci-Fi, Shounen           6.88
Action, Adventure, Cars, Mecha, Sci-Fi, Shounen, Sports    6.46
Action, Adventure, Cars, Sci-Fi                            6.86
                                                           ... 
Super Power, Supernatural, Vampire                         4.76
Supernatural                                               6.40
Thriller                                                   6.51
Vampire                                                    4.24
Yaoi                                                       6.08
Length: 3260, dtype: float64

In [ ]:
anime_genre_sim = 1 - pairwise_distances( anime_genre.values,metric='cosine')

In [ ]:
anime_genre_sim

array([[1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 1., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]])

In [ ]:
G_data=pd.DataFrame(anime_genre_sim)
G_data

0      1      2      3      4      5      6      7      8      9      \
0        1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1        0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2        0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3        0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0   
4        0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
12205    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
12206    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
12207    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
12208    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
12209    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

       ...  12200  12201  12202  12203  12204  12205  12206  12207  12208  \
0      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2      ...    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0   
3      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
12205  ...    0.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0   
12206  ...    0.0    0.0    0.0    1.0    0.0    0.0    1.0    0.0    0.0   
12207  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1.0    0.0   
12208  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1.0   
12209  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

       12209  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
...      ...  
12205    0.0  
12206    0.0  
12207    0.0  
12208    0.0  
12209    1.0  

[12210 rows x 12210 columns]

In [ ]:
G_data.index= anime1.index.unique()
G_data.columns= anime1.index.unique()

In [ ]:
G_data

anime_id  1      5      6      7      8      15     16     17     18     \
anime_id                                                                  
1           1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5           0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
6           0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0    0.0   
7           0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   
8           0.0    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0   
...         ...    ...    ...    ...    ...    ...    ...    ...    ...   
34514       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34519       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34522       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34525       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34527       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  19     ...  34497  34498  34502  34503  34506  34514  34519  34522  \
anime_id         ...                                                           
1           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
6           0.0  ...    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   
7           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
8           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...         ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
34514       0.0  ...    0.0    0.0    0.0    0.0    0.0    1.0    0.0    0.0   
34519       0.0  ...    0.0    0.0    0.0    1.0    0.0    0.0    1.0    0.0   
34522       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    1.0   
34525       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34527       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  34525  34527  
anime_id                
1           0.0    0.0  
5           0.0    0.0  
6           0.0    0.0  
7           0.0    0.0  
8           0.0    0.0  
...         ...    ...  
34514       0.0    0.0  
34519       0.0    0.0  
34522       0.0    0.0  
34525       1.0    0.0  
34527       0.0    1.0  

[12210 rows x 12210 columns]

In [ ]:
np.fill_diagonal(anime_genre_sim,val=0)
G_data


anime_id  1      5      6      7      8      15     16     17     18     \
anime_id                                                                  
1           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
6           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
7           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
8           0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...         ...    ...    ...    ...    ...    ...    ...    ...    ...   
34514       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34519       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34522       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34525       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34527       0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  19     ...  34497  34498  34502  34503  34506  34514  34519  34522  \
anime_id         ...                                                           
1           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
5           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
6           0.0  ...    0.0    0.0    1.0    0.0    0.0    0.0    0.0    0.0   
7           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
8           0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...         ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
34514       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34519       0.0  ...    0.0    0.0    0.0    1.0    0.0    0.0    0.0    0.0   
34522       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34525       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
34527       0.0  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

anime_id  34525  34527  
anime_id                
1           0.0    0.0  
5           0.0    0.0  
6           0.0    0.0  
7           0.0    0.0  
8           0.0    0.0  
...         ...    ...  
34514       0.0    0.0  
34519       0.0    0.0  
34522       0.0    0.0  
34525       0.0    0.0  
34527       0.0    0.0  

[12210 rows x 12210 columns]

In [ ]:
G_data.idxmax(axis=1)[:10]

anime_id
1      4037
5         1
6      2097
7         1
8       613
15      995
16        1
17    30652
18      185
19        1
dtype: int64

In [ ]:
anime[(anime['anime_id']==15)| (anime['anime_id']==4037)]

anime_id                             name  \
433         15                     Eyeshield 21   
1465      4037  Cowboy Bebop: Yose Atsume Blues   

                                                genre     type episodes  \
433                   Action, Comedy, Shounen, Sports       TV      145   
1465  Action, Adventure, Comedy, Drama, Sci-Fi, Space  Special        1   

      rating  members  
433     8.08    83648  
1465    7.53    19565

In [ ]:
anime[(anime['anime_id']== 254)| (anime['anime_id']==21)]

anime_id                                    name  \
74         21                               One Piece   
590       254  Jungle wa Itsumo Hare nochi Guu Deluxe   

                                                 genre type episodes  rating  \
74   Action, Adventure, Comedy, Drama, Fantasy, Sho...   TV  Unknown    8.58   
590                              Comedy, Slice of Life  OVA        6    7.95   

     members  
74    504862  
590     8847

In [ ]:
anime[(anime['anime_id']==1565)| (anime['anime_id']==1137)]

anime_id                         name  \
2558      1565  Pokemon Diamond &amp; Pearl   
2958      1137                    Mushrambo   

                                           genre type episodes  rating  \
2558    Action, Adventure, Comedy, Fantasy, Kids   TV      191    7.23   
2958  Adventure, Drama, Fantasy, Sci-Fi, Shounen   TV       32    7.12   

      members  
2558    48203  
2958     8722

In [ ]:
anime.head()

anime_id                              name  \
0     32281                    Kimi no Na wa.   
1      5114  Fullmetal Alchemist: Brotherhood   
2     28977                          Gintama°   
3      9253                       Steins;Gate   
4      9969                     Gintama&#039;   

                                               genre   type episodes  rating  \
0               Drama, Romance, School, Supernatural  Movie        1    9.37   
1  Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64    9.26   
2  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.25   
3                                   Sci-Fi, Thriller     TV       24    9.17   
4  Action, Comedy, Historical, Parody, Samurai, S...     TV       51    9.16   

   members  
0   200630  
1   793665  
2   114262  
3   673572  
4   151266

# Recommendation system

In [ ]:
anime_final = anime.set_index('anime_id')
anime_final

name  \
anime_id                                                      
32281                                        Kimi no Na wa.   
5114                       Fullmetal Alchemist: Brotherhood   
28977                                              Gintama°   
9253                                            Steins;Gate   
9969                                          Gintama&#039;   
...                                                     ...   
9316           Toushindai My Lover: Minami tai Mecha-Minami   
5543                                            Under World   
5621                         Violence Gekiga David no Hoshi   
6133      Violence Gekiga Shin David no Hoshi: Inma Dens...   
26081                      Yasuji no Pornorama: Yacchimae!!   

                                                      genre   type episodes  \
anime_id                                                                      
32281                  Drama, Romance, School, Supernatural  Movie        1   
5114      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
28977     Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
9253                                       Sci-Fi, Thriller     TV       24   
9969      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                     ...    ...      ...   
9316                                                 Hentai    OVA        1   
5543                                                 Hentai    OVA        1   
5621                                                 Hentai    OVA        4   
6133                                                 Hentai    OVA        1   
26081                                                Hentai  Movie        1   

          rating  members  
anime_id                   
32281       9.37   200630  
5114        9.26   793665  
28977       9.25   114262  
9253        9.17   673572  
9969        9.16   151266  
...          ...      ...  
9316        4.15      211  
5543        4.28      183  
5621        4.88      219  
6133        4.98      175  
26081       5.46      142  

[12210 rows x 6 columns]

In [ ]:
def recommend(id):
    g= anime_final[anime_final.index==id].genre.item()

    M=anime_final[anime_final.index==id].name.item()

    movies = anime_final[anime_final['genre']==g].name.values

    movies=movies.tolist()

    movies.remove(M)

    if len(movies)!=0:
        print( list(movies))

    else:
        type = anime_final['type'].loc[id]

        movies_t= anime_final[(anime_final['type']==type) & (anime_final['rating']>=6)].name.tolist()
                                                                                                #list(anime_final[anime_final['type']==type].name)
        if anime_final['name'].loc[id] in movies_t:
            movies_t.remove(anime_final['name'].loc[3455])

        print(movies_t)


In [ ]:
anime_final[anime_final.index== 515]

name  \
anime_id                                                     
515       Hajimari no Boukensha-tachi: Legend of Crystania   

                                                    genre   type episodes  \
anime_id                                                                    
515       Action, Adventure, Fantasy, Magic, Supernatural  Movie        1   

          rating  members  
anime_id                   
515         6.15     2893

In [ ]:
recommend(515)

['Juuni Kokuki', 'Legend of Crystania OVA']


In [ ]:
anime_final[anime_final.index==10000]

name  \
anime_id                                           
10000     Esper Mami Special: My Angel Mami-chan   

                                      genre     type episodes  rating  members  
anime_id                                                                        
10000     Comedy, Super Power, Supernatural  Special        1    6.52      148

In [ ]:
recommend(1000)

['Space Pirate Captain Herlock: Outside Legend - The Endless Odyssey', 'Harlock Saga: Nibelung no Yubiwa']
